packages/core/http/core-http-server-internal/src/lifecycle_handlers.ts

```diff
@@ -39,6 +40,27 @@ export const createXsrfPostAuthHandler = (config: HttpConfig): OnPostAuthHandler
   };
 };
 
+export const createRestrictInternalRoutesPostAuthHandler = (
+  config: HttpConfig
+): OnPostAuthHandler => {
+  const isRestrictionEnabled = config.restrictInternalApis;
+
+  return (request, response, toolkit) => {
+    const isInternalRoute = request.route.options.access === 'internal';
+
+    // only check if the header is present, not it's content.
+    const hasInternalKibanaRequestHeader = X_ELASTIC_INTERNAL_ORIGIN_REQUEST in request.headers;
+
+    if (isRestrictionEnabled && isInternalRoute && !hasInternalKibanaRequestHeader) {
+      // throw 400
+      return response.badRequest({
+        body: `uri [${request.url}] with method [${request.route.method}] exists but is not available with the current configuration`,
```

Comment:
I like that this response does not reveal anything about why the request was not handled, and it is distinct from other bad request responses. One small iteration on the current response message:

```
...exists but is not available to this request.
```

I have one alternative approach for consideration: this `access` control is not a security feature, rather a UX feature. I still wonder whether we could think of it as access control but specific to application logic. The `403 Forbidden` status code seems to communicate this idea [based on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403). In that case we could just return `403 Forbidden` without further details.

Let me know what you think!

Reply:
> In that case we could just return 403 Forbidden without further details.

I also thought 403 at first and changed my mind to be more consistent with ES (As Luke pointed out [here](https://github.com/elastic/kibana/issues/151940#issuecomment-1446580700)).